pkg/metricsadapter/provider/custommetrics.go

```diff
@@ -3,33 +3,353 @@ package provider
+	"strings"
+	"sync"
 
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/discovery"
+	"k8s.io/klog/v2"
 	"k8s.io/metrics/pkg/apis/custom_metrics"
+	custommetricsv1beta2 "k8s.io/metrics/pkg/apis/custom_metrics/v1beta2"
+	custommetricsclient "k8s.io/metrics/pkg/client/custom_metrics"
+	custommetricsschema "k8s.io/metrics/pkg/client/custom_metrics/scheme"
 	"sigs.k8s.io/custom-metrics-apiserver/pkg/provider"
+
+	clusterlister "github.com/karmada-io/karmada/pkg/generated/listers/cluster/v1alpha1"
+	"github.com/karmada-io/karmada/pkg/metricsadapter/multiclient"
+)
+
+var (
+	versionConverter = custommetricsclient.NewMetricConverter()
 )
 
 // CustomMetricsProvider is a custom metrics provider
 type CustomMetricsProvider struct {
+	// multiClusterDiscovery returns a discovery client for member cluster apiserver
+	multiClusterDiscovery multiclient.MultiClusterDiscoveryInterface
+	clusterLister         clusterlister.ClusterLister
 }
 
 // MakeCustomMetricsProvider creates a new custom metrics provider
-func MakeCustomMetricsProvider() *CustomMetricsProvider {
-	return &CustomMetricsProvider{}
+func MakeCustomMetricsProvider(clusterLister clusterlister.ClusterLister, multiClusterDiscovery multiclient.MultiClusterDiscoveryInterface) *CustomMetricsProvider {
+	return &CustomMetricsProvider{
+		clusterLister:         clusterLister,
+		multiClusterDiscovery: multiClusterDiscovery,
+	}
 }
 
 // GetMetricByName will query metrics by name from member clusters and return the result
 func (c *CustomMetricsProvider) GetMetricByName(ctx context.Context, name types.NamespacedName, info provider.CustomMetricInfo, metricSelector labels.Selector) (*custom_metrics.MetricValue, error) {
-	return nil, fmt.Errorf("karmada-metrics-adapter still not implement it")
+	clusters, err := c.clusterLister.List(labels.Everything())
+	if err != nil {
+		klog.Errorf("Failed to list clusters: %v", err)
+		return nil, err
+	}
+	metricValueList := &custom_metrics.MetricValueList{}
+	metricsChanel := make(chan *custom_metrics.MetricValueList)
+	wg := sync.WaitGroup{}
+	for _, cluster := range clusters {
+		wg.Add(1)
+		go func(clusterName string) {
+			defer wg.Done()
+			metrics, err := c.getMetricByName(ctx, clusterName, name, info, metricSelector)
+			if err != nil {
+				klog.Warningf("query %s's %s metric from cluster %s failed, err: %+v", info.GroupResource.String(), info.Metric, clusterName, err)
+				return
+			}
+			metricsChanel <- metrics
+		}(cluster.Name)
+	}
+	go func() {
+		wg.Wait()
+		close(metricsChanel)
+	}()
+	for {
+		metrics, ok := <-metricsChanel
+		if !ok {
+			break
+		}
+		metricValueList.Items = append(metricValueList.Items, metrics.Items...)
+	}
+	var metrics *custom_metrics.MetricValue
+	// TODO(chaunceyjiang) The MetricValue items need to be sorted.
```

Comment:
Why MetricValue items need to be sorted?

Reply:
Because even if they are the same pod, their timestamps may be different. 



<img width="489" alt="image" src="https://github.com/karmada-io/karmada/assets/17962021/1a4b8a19-b21d-485a-9848-4ab739a83d4a">


So it may be necessary to sort in order to maintain the stability of the returned results.